<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import requests
from bs4 import BeautifulSoup

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [8]:
# текст запроса
query_3_1 = f'''select count(id)
           from public.vacancies
        '''

In [34]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
print(f"Количество вакансий в базе: {df['count'][0]}")

Количество вакансий в базе: 49197


C:\Users\romep\AppData\Local\Temp\ipykernel_16592\2829240399.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [20]:
# текст запроса
query_3_2 = f'''select count(id)
           from public.employers
        '''

In [31]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
print(f"Количество работодателей в базе: {df['count'][0]}")

Количество работодателей в базе: 23501


C:\Users\romep\AppData\Local\Temp\ipykernel_16592\4257310605.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [25]:
# текст запроса
query_3_3 = f'''select count(id)
           from public.areas
        '''

In [32]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
print(f"Количество регионов в базе: {df['count'][0]}")

Количество регионов в базе: 1362


C:\Users\romep\AppData\Local\Temp\ipykernel_16592\338425011.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [41]:
# текст запроса
query_3_4 = f'''select count(id)
           from public.industries
        '''

In [33]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
print(f"Количество сфер деятельности в базе: {df['count'][0]}")

Количество сфер деятельности в базе: 294


C:\Users\romep\AppData\Local\Temp\ipykernel_16592\1229991471.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


# Выводы по предварительному анализу данных:
* Количество вакансий в базе: 49197;
* Количество работодателей в базе: 23501;
* Количество регионов в базе: 1362;
* Количество сфер деятельности в базе: 294.

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [42]:
# текст запроса
query_4_1 = f'''select a.name "Название региона", count(v.id) "Количество вакансий"
           from public.vacancies v
           join public.areas a on a.id = v.area_id
           group by "Название региона"
           order by "Количество вакансий" desc
           limit 5
        '''

In [43]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_16592\234815181.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


Название региона  Количество вакансий
0           Москва                 5333
1  Санкт-Петербург                 2851
2            Минск                 2112
3      Новосибирск                 2006
4           Алматы                 1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [46]:
# текст запроса
query_4_2 = f'''select count(id)
           from public.vacancies
           where salary_from is not null or salary_to is not null
        '''

In [47]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
print(f"Количество вакансий с заполненной зарплатой: {df['count'][0]}")

Количество вакансий с заполненной зарплатой: 24073


C:\Users\romep\AppData\Local\Temp\ipykernel_16592\3589055623.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [58]:
# текст запроса
query_4_3 = f'''select avg(salary_from) avg_from, avg(salary_to) avg_to 
           from public.vacancies
        '''

In [64]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
print(f"Среднее нижней границы заработной платы: {round(df['avg_from'][0])}")
print(f"Среднее верхней границы заработной платы: {round(df['avg_to'][0])}")

Среднее нижней границы заработной платы: 71065
Среднее верхней границы заработной платы: 110537


C:\Users\romep\AppData\Local\Temp\ipykernel_16592\4191545187.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [67]:
# текст запроса
query_4_4 = f'''select schedule "Тип рабочего графика",
           employment "Тип трудоустройства",
           count(id) "Количество вакансий"
           from public.vacancies
           group by "Тип рабочего графика", "Тип трудоустройства"
           order by "Количество вакансий" desc
           limit 5
        '''

In [68]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_16592\3274844153.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Тип рабочего графика  Тип трудоустройства  Количество вакансий
0          Полный день     Полная занятость                35367
1     Удаленная работа     Полная занятость                 7802
2        Гибкий график     Полная занятость                 1593
3     Удаленная работа  Частичная занятость                 1312
4       Сменный график     Полная занятость                  940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [69]:
# текст запроса
query_4_5 = f'''select experience "Требуемый опыт", count(id) "Количество вакансий"
           from public.vacancies
           group by "Требуемый опыт"
           order by "Количество вакансий" asc
           limit 5
        '''

In [70]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_16592\969833449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Требуемый опыт  Количество вакансий
0         Более 6 лет                 1337
1           Нет опыта                 7197
2       От 3 до 6 лет                14511
3  От 1 года до 3 лет                26152

# Выводы по детальному анализу вакансий:

* Пятёрка лидеров по количеству вакансий: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы;
* Количество вакансий с заполненной зарплатой: 24073;
* Среднее нижней границы заработной платы: 71065;
* Среднее верхней границы заработной платы: 110537;
* Самые популярный тип рабочего графика и тип трудоустройства: полный день - полная занятость;
* Требуемый опыт работы в порядке возрастания количества вакансий: Более 6 лет — Нет опыта — От 3 до 6 лет — От 1 года до 3 лет

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [139]:
# текст запроса
query_5_1 = f'''select e.name "Работодатель", count(v.id) "Количество вакансий"
           from public.vacancies v
           join public.employers e on e.id = v.employer_id
           group by "Работодатель"
           order by "Количество вакансий" desc
           limit 5
        '''

In [140]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_16592\1127888558.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Работодатель  Количество вакансий
0         Яндекс                 1933
1     Ростелеком                  491
2       Тинькофф                  444
3           СБЕР                  428
4  Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [156]:
# текст запроса
query_5_2 = f'''select a.name "Регион", 
           count(e.id) "Количество работодателей",
           count(v.id) "Количество вакансий"
           from public.areas a
           left join public.vacancies v on a.id = v.area_id
           left join public.employers e on a.id = e.area
           where v.id is null
           group by a.id
           order by "Количество работодателей" desc
           limit 5
        '''

In [157]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_16592\2002098862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


Регион  Количество работодателей  Количество вакансий
0              Россия                       410                    0
1           Казахстан                       207                    0
2  Московская область                        75                    0
3  Краснодарский край                        19                    0
4            Беларусь                        18                    0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [23]:
# текст запроса
query_5_3 = f'''select e.name "Работодатель", 
           count(distinct area_id) "Количество регионов"
           from public.employers e
           left join public.vacancies v on v.employer_id = e.id
           group by e.id
           order by "Количество регионов" desc
           limit 5
        '''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_13912\3953667891.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Работодатель  Количество регионов
0                  Яндекс                  181
1              Ростелеком                  152
2              Спецремонт                  116
3  Поляков Денис Иванович                   88
4               ООО ЕФИН                    71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [4]:
# текст запроса
query_5_4 = f'''select count(e.id) 
           from public.employers e
           left join public.employers_industries ei on ei.employer_id = e.id
           where ei.industry_id is null
        '''

In [5]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
print(f"Количество работодателей с неуказанной сферой деятельности: {df['count'][0]}")

Количество работодателей с неуказанной сферой деятельности: 8419


C:\Users\romep\AppData\Local\Temp\ipykernel_15764\2434766366.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [15]:
# текст запроса
query_5_5 = f'''select e.name "Название компании",
           count(ei.industry_id) "Количество сфер деятельности" 
           from public.employers e
           left join public.employers_industries ei on ei.employer_id = e.id
           group by e.id
           having count(ei.industry_id) = 4
           order by "Название компании" asc
        '''

In [16]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_15764\3277383447.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Название компании  Количество сфер деятельности
0         101 Интернет                             4
1             21vek.by                             4
2                 2ГИС                             4
3                   2К                             4
4          4 пикселя +                             4
...                ...                           ...
1133             ЮРИОН                             4
1134   ЮТИП Технологии                             4
1135            ЯКласс                             4
1136             ЯрНео                             4
1137     Яшин&Партнёры                             4

[1138 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [21]:
# текст запроса
query_5_6 = f'''select count(ei.employer_id) 
           from  public.employers_industries ei
           join public.industries i on i.id = ei.industry_id
           where i.name = 'Разработка программного обеспечения'
        '''

In [22]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
print(f"Количество работодателей со сферой деятельности «Разработка программного обеспечения»: {df['count'][0]}")

Количество работодателей со сферой деятельности «Разработка программного обеспечения»: 3553


C:\Users\romep\AppData\Local\Temp\ipykernel_15764\2928326827.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [83]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров'
response = requests.get(url) 
page = BeautifulSoup(response.text, 'html.parser')
links = page.find_all('a') 
city = []
for link in links[47:63]:
    city.append(link.text)
#print([link.text for link in links[47:63]])
city = tuple(city)
print(city)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [122]:
# текст запроса

query_5_7 = f'''with x as (
         select a.name "Регион миллионник", count(v.id) "Количество вакансий"
         from  public.vacancies v
         join public.employers e on e.id = v.employer_id
         join public.areas a on a.id = v.area_id
         where e.name = 'Яндекс'
         group by "Регион миллионник"
         having a.name in {city})
         
         select x."Регион миллионник", x."Количество вакансий"
         from x
         union
         select 'total', sum(x."Количество вакансий")
         from x
         order by 2
        '''

In [123]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_15764\2643464226.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


Регион миллионник  Количество вакансий
0               Омск                 21.0
1          Челябинск                 22.0
2         Красноярск                 23.0
3          Волгоград                 24.0
4              Пермь                 25.0
5             Казань                 25.0
6     Ростов-на-Дону                 25.0
7                Уфа                 26.0
8             Самара                 26.0
9          Краснодар                 30.0
10           Воронеж                 32.0
11       Новосибирск                 35.0
12   Нижний Новгород                 36.0
13      Екатеринбург                 39.0
14   Санкт-Петербург                 42.0
15            Москва                 54.0
16             total                485.0

# Выводы по анализу работодателей:

* Пятёрка работодателей по количеству вакансий: Яндекс, Ростелеком, Тинькофф, СБЕР, Газпром нефть;
* Наибольшее количество работодателей представлено в стране РФ;
* Максимальное число вакансий в регионах публикует компания Яндекс;
* Количество работодателей с неуказанной сферой деятельности: 8419;
* Количество работодателей, у которых указано четыре сферы деятельности: 1138;
* Количество работодателей со сферой деятельности «Разработка программного обеспечения»: 3553;
* Общее число вакансий компании Яндекс в городах-миллионниках: 485.

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [4]:
# текст запроса
query_6_1 = f'''select count(id) 
           from  public.vacancies
           where lower(name) like '%data%' or lower(name) like '%данн%'
        '''

In [174]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
print(f"Отношение к данным имеет: {df['count'][0]} вакансия")

Отношение к данным имеет: 1771 вакансия


C:\Users\romep\AppData\Local\Temp\ipykernel_15072\1050086887.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [114]:
# текст запроса
query_6_2= f'''with x as (
         select *
         from  public.vacancies
         where name ilike '%data scientist%' or name ilike '%data science%' or
         name ilike '%исследователь данных%' or (name ilike '%ML%' and name not ilike '%HTML%') or
         name ilike '%machine learning%' or name ilike '%машинн%обучен%'
         )
         
         select count(x.id) 
         from  x
         where x.name ilike '%junior%' or x.experience = 'Нет опыта' or x.schedule = 'Стажировка'
        '''

In [115]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
print(f"Вакансий специалистов уровня junior: {df['count'][0]}")

C:\Users\romep\AppData\Local\Temp\ipykernel_15072\2079563719.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


Вакансий специалистов уровня junior: 51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [116]:
# текст запроса
query_6_3= f'''with x as (
         select *
         from  public.vacancies
         where name ilike '%data scientist%' or name ilike '%data science%' or
         name ilike '%исследователь данных%' or (name ilike '%ML%' and name not ilike '%HTML%') or
         name ilike '%machine learning%' or name ilike '%машинн%обучен%'
         )
         
         select count(x.id) 
         from  x
         where x.key_skills ilike '%SQL%' or x.key_skills ilike '%postgres%' 
        '''

In [117]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
print(f"Вакансий специалистов с навыками SQL или Postgres: {df['count'][0]}")

C:\Users\romep\AppData\Local\Temp\ipykernel_15072\3855912623.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


Вакансий специалистов с навыками SQL или Postgres: 229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [118]:
# текст запроса
query_6_4= f'''with x as (
         select *
         from  public.vacancies
         where name ilike '%data scientist%' or name ilike '%data science%' or
         name ilike '%исследователь данных%' or (name ilike '%ML%' and name not ilike '%HTML%') or
         name ilike '%machine learning%' or name ilike '%машинн%обучен%'
         )
         
         select count(x.id) 
         from  x
         where x.key_skills ilike '%Python%'
        '''

In [119]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
print(f"Вакансий специалистов с навыками Python: {df['count'][0]}")

C:\Users\romep\AppData\Local\Temp\ipykernel_15072\2409390534.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


Вакансий специалистов с навыками Python: 357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [120]:
# текст запроса
query_6_5= f'''with x as (
         select *
         from  public.vacancies
         where name ilike '%data scientist%' or name ilike '%data science%' or
         name ilike '%исследователь данных%' or (name like '%ML%' and name not like '%HTML%') or
         name ilike '%machine learning%' or name ilike '%машинн%обучен%'
         )
         
         select x.key_skills 
         from  x
         where x.key_skills is not null
        '''

In [145]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df['key_skills'] = df['key_skills'].apply(lambda x: len(str(x).split('\t')))
print(f"Специалисты по DS в среднем указывают: {round(df['key_skills'].mean(),2)} навыков")

C:\Users\romep\AppData\Local\Temp\ipykernel_15072\1513109865.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Специалисты по DS в среднем указывают: 6.41 навыков


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [172]:
# текст запроса
query_6_6= f'''with x as (
         select *
         from  public.vacancies
         where name ilike '%data scientist%' or name ilike '%data science%' or
         name ilike '%исследователь данных%' or (name like '%ML%' and name not like '%HTML%') or
         name ilike '%machine learning%' or name ilike '%машинн%обучен%'
         )
         
         select x.experience "Опыт", round(avg(coalesce((x.salary_to + x.salary_from)/2, x.salary_from, x.salary_to))) "Средняя зарплата"
         from  x
         where x.salary_from is not null or x.salary_to is not null
         group by experience
        '''


In [173]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_15072\3373444971.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


Опыт  Средняя зарплата
0       От 3 до 6 лет          243115.0
1  От 1 года до 3 лет          139675.0
2           Нет опыта           74643.0

***

# Выводы по предметному анализу:

* Отношение к данным имеет: 1771 вакансия;
* Вакансий специалистов уровня junior: 51;
* Вакансий специалистов с навыками SQL или Postgres: 229;
* Вакансий специалистов с навыками Python: 357;
* Специалисты по DS в среднем указывают 6 ключевых навыков;
* Дата-сайентист с опытом работы от 3 до 6 лет в среднем может рассчитывать на зарплату: 243115 рублей.

# Общий вывод по проекту

# Подведем итог исследования:

- На сайте hh.ru большинство работодателей представлено из крупнейших городов России и стран СНГ. Большинстов вакансий публикуют  такие крупные холдинги как Яндекс, Ростелеком, Тинькофф, СБЕР, Газпром нефть. Максимальное число вакансий в регионах публикует компания Яндекс и по логике также по нисходящей должны были быть представлены холдинги, упомянутые ранее, но представлены менее известные компании (Спецремонт -116; Поляков Денис Иванович - 88; ООО ЕФИН - 71).
- Наиболее популярный тип графика полный день - полная занятость, данные компании не настолько крупные, что их филиалы, если они вообще есть, присутствовали бы в столь большом количестве регионов и возможности данных компаний не позволили бы им предоставлять удаленный режим работы;
- Также нам известно, что из всех работодателей (23501) около трети имеет не заполненную сферу деятельности, возможно данные компании занимаются мошеннической деятельностью и не верно указывают информацию, среди которых возможно присутствуют (Спецремонт; Поляков Денис Иванович; ООО ЕФИН). В связи с этим проведем дополнительное исследование.



In [57]:
# текст запроса
query_7_1 = f'''select e.name "Наименование компании", i.name "Наименование сферы деятельности" 
           from public.employers e
           left join public.employers_industries ei on ei.employer_id = e.id
           left join public.industries i on ei.industry_id = i.id
           where e.name = 'Спецремонт' or e.name = 'Поляков Денис Иванович' or e.name ilike '%ЕФИН%'
        '''

In [58]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
display(df)

C:\Users\romep\AppData\Local\Temp\ipykernel_13912\2101136656.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


Наименование компании                    Наименование сферы деятельности
0              Спецремонт                                               None
1  Поляков Денис Иванович                                               None
2               ООО ЕФИН                                                None
3         Рефиниш Автолак  Лакокрасочная продукция, сырье (продвижение, о...
4         Рефиниш Автолак      Лакокрасочная продукция, сырье (производство)

- Мы видим, что у данных организаций действительно не заполнены поля с наименованием сферы деятельности. Если данные компании действительно являются мошенническими, то это существенно искажают картину по вакансиям. В связи с этим специалистам по безопасности hh.ru необходимо произвести дополнительную проверку, чтобы удостовериться в надежности данных организаций. 
- Подводя итог исследованию, можно сказать, что лишь 1771 вакансия из 49197 может быть использована в качестве основы, для последующих этапов в процессе создания модели машинного обучения, которая будет рекомендовать вакансии клиентам агентства, претендующим на позицию Data Scientist. 